xblock-external-ui: Add XBlock API call to render XBlock views

antoviaque opened the pull request:
@cpennington @nedbat @ormsbee Can you have a quick look to check that I'm on the right path with this? It seem to work well, rendering either individual blocks or verticals containing both XBlocks and XModules. If that's the correct approach, I'll flesh it out more, add tests and start working on the client side, but I could use a sanity check : ) Thanks!

antoviaque commented:
@singingwolfboy Rebased, hopefully we should be good now : )

ormsbee commented on lms/urls.py:
```diff
@@ -217,6 +217,9 @@
             name='xblock_handler'),
+        url(r'^courses/{course_key}/xblock/{usage_key}/view/(?P<view_name>[^/]*)$'.format(course_key=settings.COURSE_ID_PATTERN, usage_key=settings.USAGE_ID_PATTERN),
+            'courseware.module_render.xblock_view',
+            name='xblock_view'),
```
Does the URL exist regardless of whether we've enabled the settings to allow it to work properly? What happens if you try to invoke it?

ormsbee commented:
My one request is that we have a flag that will enable/disable the endpoint altogether (I don't care if it actually changes the urls.py or just returns immediately from the view).

Otherwise, I'm :+1:

singingwolfboy commented on lms/djangoapps/courseware/module_render.py:
```diff
@@ -912,6 +933,52 @@ def _invoke_xblock_handler(request, course_id, usage_id, handler, suffix, user):
 
+def hash_resource(resource):
+    """
+    Hash a :class:`xblock.fragment.FragmentResource
```
Nitpick: this needs an ending `

singingwolfboy commented:
:+1:

sarina commented:
I'm asking test eng about the workers failing to start....

jzoldak commented:
test this please

sarina commented:
@antoviaque - can you fix the quality issues? Otherwise build is passing and looks good to go w/ 2 thumbs.

antoviaque commented:
@sarina I fixed the the quality issues and rebased, but I'm now getting failures in acceptance tests. I've tried re-running it a couple of times without success. I'll try again tomorrow in case this is an issue with the current master or jenkins workers.

antoviaque commented:
@sarina Seem to have worked better - green build! : )

How do you want to proceed for the merge? Should I merge it myself, or should it be someone else given that it's an OSPR ticket?

sarina commented:
@antoviaque - I don't really have a strong opinion. You can merge it yourself if you are able to verify the changes on stage and be on hand for any potential issues that arise. If not, one of your reviewers should merge it.

antoviaque commented:
@sarina Yup, I already do the stage verification thing for other feature merges when I merge SOL tickets (and the current PR is now behind a feature flag so there won't be much to do), so that's fine. And I am of course happy to help if any issue arises - if I wasn't around when an issue would happen (for example if it's late during the day, since I'm currently on a European timezone), others from OpenCraft can help too (Braden and Jonathan for example are on the PST timezone). If that sounds good to you, I'll go ahead and merge.

Thanks to you for the help in getting that PR merged btw - the longest-lasting PR I would guess : )

sarina commented:
@antoviaque i think there are older :cry: 

but yes - go ahead and merge, I think!

antoviaque commented:
@sarina Sounds good - thanks! And good luck with caretaking the other ancient PRs : )
